web3.js/src/connection.ts

```diff
@@ -2865,39 +2909,72 @@ export class Connection {
       }
     });
 
-    let timeoutMs = this._confirmTransactionInitialTimeout || 60 * 1000;
-    switch (subscriptionCommitment) {
-      case 'processed':
-      case 'recent':
-      case 'single':
-      case 'confirmed':
-      case 'singleGossip': {
-        timeoutMs = this._confirmTransactionInitialTimeout || 30 * 1000;
-        break;
+    const checkBlockHeight = async () => {
+      const blockHeight = await this.getBlockHeight(commitment);
+      return blockHeight;
+    };
+
+    const expireTx = new Promise((resolve, reject) => {
+      if (typeof strategy === 'string') {
+        console.log('timeout');
+        let timeoutMs = this._confirmTransactionInitialTimeout || 60 * 1000;
+        switch (subscriptionCommitment) {
+          case 'processed':
+          case 'recent':
+          case 'single':
+          case 'confirmed':
+          case 'singleGossip': {
+            timeoutMs = this._confirmTransactionInitialTimeout || 30 * 1000;
+            break;
+          }
+          // exhaust enums to ensure full coverage
+          case 'finalized':
+          case 'max':
+          case 'root':
+        }
+        if (done) clearTimeout();
+        setTimeout(() => resolve(TransactionStatus.EXPIRED), timeoutMs);
+      } else if (Object.prototype.hasOwnProperty.call(strategy, 'signature')) {
+        let config = strategy as BlockheightBasedTransactionConfimationStrategy;
+        (async () => {
+          try {
+            let currentBlockHeight = await checkBlockHeight();
+            console.log(currentBlockHeight);
+            console.log(config.lastValidBlockHeight);
+            if (done) return;
+            while (currentBlockHeight <= config.lastValidBlockHeight) {
+              await sleep(1000);
+              if (done) return;
+              currentBlockHeight = await checkBlockHeight();
+              if (done) return;
+            }
+            resolve(TransactionStatus.EXPIRED);
+          } catch (error) {
+            reject(error);
+          }
+        })();
       }
-      // exhaust enums to ensure full coverage
-      case 'finalized':
-      case 'max':
-      case 'root':
-    }
+    });
 
     try {
-      await promiseTimeout(confirmPromise, timeoutMs);
+      const outcome = await Promise.race([confirmTx, expireTx]);
+      switch (outcome) {
+        case 'CONFIRMED':
+          console.log('Transaction has been confirmed');
+          break;
+        case 'EXPIRED':
+          console.log('Transaction has expired');
+          break;
+      }
     } finally {
       if (subscriptionId) {
         this.removeSignatureListener(subscriptionId);
       }
     }
```

Comment:
This should be the last line in your function. The `switch` should always return from one of the branches.